python/_tskitmodule.c

```diff
@@ -11295,6 +11301,221 @@ static PyTypeObject VariantGeneratorType = {
+        tsk_variant_free(self->variant);
+        PyMem_Free(self->variant);
+        self->variant = NULL;
+    }
+    Py_XDECREF(self->tree_sequence);
+    Py_TYPE(self)->tp_free((PyObject *) self);
+}
+
+static int
+Variant_init(Variant *self, PyObject *args, PyObject *kwds)
+{
+    int ret = -1;
+    int err;
+    static char *kwlist[]
+        = { "tree_sequence", "samples", "isolated_as_missing", "alleles", NULL };
+    TreeSequence *tree_sequence = NULL;
+    PyObject *samples_input = Py_None;
+    PyObject *py_alleles = Py_None;
+    PyArrayObject *samples_array = NULL;
+    tsk_id_t *samples = NULL;
+    tsk_size_t num_samples = 0;
+    int isolated_as_missing = 1;
+    const char **alleles = NULL;
+    npy_intp *shape;
+    tsk_flags_t options = 0;
+
+    self->variant = NULL;
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O!|OiO", kwlist, &TreeSequenceType,
+            &tree_sequence, &samples_input, &isolated_as_missing, &py_alleles)) {
+        goto out;
+    }
+    if (!isolated_as_missing) {
+        options |= TSK_ISOLATED_NOT_MISSING;
+    }
+    /* tsk_variant_t holds a reference to the tree sequence so we must too*/
+    self->tree_sequence = tree_sequence;
+    Py_INCREF(self->tree_sequence);
+    if (TreeSequence_check_state(self->tree_sequence) != 0) {
+        goto out;
+    }
+    if (samples_input != Py_None) {
+        samples_array = (PyArrayObject *) PyArray_FROMANY(
+            samples_input, NPY_INT32, 1, 1, NPY_ARRAY_IN_ARRAY);
+        if (samples_array == NULL) {
+            goto out;
+        }
+        shape = PyArray_DIMS(samples_array);
+        num_samples = (tsk_size_t) shape[0];
+        samples = PyArray_DATA(samples_array);
+    }
+    if (py_alleles != Py_None) {
+        alleles = parse_allele_list(py_alleles);
+        if (alleles == NULL) {
+            goto out;
+        }
+    }
+    self->variant = PyMem_Malloc(sizeof(tsk_vargen_t));
+    if (self->variant == NULL) {
+        PyErr_NoMemory();
+        goto out;
+    }
+    /* Note: the variant currently takes a copy of the samples list. If we wanted
+     * to avoid this we would INCREF the samples array above and keep a reference
+     * to in the object struct */
+    err = tsk_variant_init(self->variant, self->tree_sequence->tree_sequence, samples,
+        num_samples, alleles, options);
+    if (err != 0) {
+        handle_library_error(err);
+        goto out;
+    }
+    ret = 0;
+out:
+    PyMem_Free(alleles);
+    Py_XDECREF(samples_array);
+    return ret;
+}
+
+static PyObject *
+Variant_decode(Variant *self, PyObject *args, PyObject *kwds)
+{
+    int err;
+    PyObject *ret = NULL;
+    tsk_id_t site_id;
+    static char *kwlist[] = { "site", NULL };
+
+    if (Variant_check_state(self) != 0) {
+        goto out;
+    }
+    if (!PyArg_ParseTupleAndKeywords(
+            args, kwds, "O&", kwlist, &tsk_id_converter, &site_id)) {
+        goto out;
+    }
+    err = tsk_variant_decode(self->variant, site_id, 0);
```

Comment:
Should get test coverage on this - won't be covered by vargen as you can't specify the ID.

Easy to do, just pass in a bad site ID.